hc5/src/main/java/feign/hc5/AsyncApacheHttp5Client.java

```diff
@@ -0,0 +1,183 @@
+/**
+ * Copyright 2012-2020 The Feign Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package feign.hc5;
+
+import org.apache.hc.client5.http.async.methods.SimpleHttpRequest;
+import org.apache.hc.client5.http.async.methods.SimpleHttpResponse;
+import org.apache.hc.client5.http.config.Configurable;
+import org.apache.hc.client5.http.config.RequestConfig;
+import org.apache.hc.client5.http.impl.async.CloseableHttpAsyncClient;
+import org.apache.hc.client5.http.impl.async.HttpAsyncClients;
+import org.apache.hc.client5.http.protocol.HttpClientContext;
+import org.apache.hc.core5.concurrent.FutureCallback;
+import org.apache.hc.core5.http.ContentType;
+import org.apache.hc.core5.http.Header;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import feign.*;
+import feign.Request.Options;
+
+/**
+ * This module directs Feign's http requests to Apache's
+ * <a href="https://hc.apache.org/httpcomponents-client-5.0.x/index.html">HttpClient 5</a>. Ex.
+ *
+ * <pre>
+ * GitHub github = Feign.builder().client(new ApacheHttp5Client()).target(GitHub.class,
+ * "https://api.github.com");
+ */
+/*
+ */
+public final class AsyncApacheHttp5Client<C> implements AsyncClient<C> {
+
+  private static final String ACCEPT_HEADER_NAME = "Accept";
+
+  private final CloseableHttpAsyncClient client;
+
+  public AsyncApacheHttp5Client() {
+    this(HttpAsyncClients.custom().build());
+  }
+
+  public AsyncApacheHttp5Client(CloseableHttpAsyncClient client) {
+    this.client = client;
+    client.start();
```

Comment:
Doing client.start() here is a bit of an issue, because then Feign has responsibility for the lifecycle of the client (if it's doing start() it then needs to do close()). As well, the user may want to use that client in other places. So I would recommend moving that responsibility to the user.

Reply:
This is typically how our internal APIs are constructed.  Feign controls the lifecycle

Reply:
Well, I think what I will do is only start the client when no args constructor is invoked.

So, when feign creates the http client it will also start it. If you are using the overwritten constructor, you manage the lifecycle of `CloseableHttpAsyncClient`.

Also, will make this client closeable, to make possible for people using the no-args constructor to close it.